userspace/libsinsp/filterchecks.cpp

```diff
@@ -109,6 +110,317 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+}
+
+uint8_t* sinsp_filter_check_fspath::extract(sinsp_evt* evt, OUT uint32_t* len, bool sanitize_strings)
+{
+	*len = 0;
+	ASSERT(evt);
+
+	m_tstr = "";
+
+	if(m_path_checks.empty())
+	{
+		create_fspath_checks();
+	}
+
+	// First check the success conditions.
+	auto it = m_success_checks.find(evt->get_type());
+	if(it == m_success_checks.end())
+	{
+		return NULL;
+	}
+
+	if(!it->second->compare(evt))
+	{
+		return NULL;
+	}
+
+	std::optional<std::string> enter_param;
+
+	switch(m_field_id)
+	{
+	case TYPE_NAME:
+	case TYPE_NAMERAW:
+
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_MKDIR_X:
+		case PPME_SYSCALL_RMDIR_X:
+		case PPME_SYSCALL_UNLINK_X:
+			enter_param = evt->get_enter_evt_param("path");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		case PPME_SYSCALL_UNLINKAT_X:
+			enter_param = evt->get_enter_evt_param("name");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_fspath(evt, m_path_checks);
+		};
+
+		break;
+	case TYPE_SOURCE:
+	case TYPE_SOURCERAW:
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_LINK_X:
+		case PPME_SYSCALL_LINKAT_X:
+			enter_param = evt->get_enter_evt_param("newpath");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_fspath(evt, m_source_checks);
+		};
+		break;
+	case TYPE_TARGET:
+	case TYPE_TARGETRAW:
+
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_LINK_X:
+		case PPME_SYSCALL_LINKAT_X:
+			enter_param = evt->get_enter_evt_param("oldpath");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_fspath(evt, m_target_checks);
+		};
+
+		break;
+	}
+
+	if(m_tstr.empty() || m_tstr == "<NA>")
+	{
+		return NULL;
+	}
+
+	// For the non-raw fields, if the path is not absolute,
+	// prepend the cwd of the threadinfo to the path.
+	if((m_field_id == TYPE_NAME ||
+	    m_field_id == TYPE_SOURCE ||
+	    m_field_id == TYPE_TARGET)
+	   && m_tstr.front() != '/')
+	{
+		sinsp_threadinfo* tinfo = evt->get_thread_info();
+
+		std::filesystem::path tstr = tinfo->get_cwd() + m_tstr;
+		m_tstr = std::filesystem::absolute(tstr).lexically_normal().string();
+	}
+
+	RETURN_EXTRACT_STRING(m_tstr);
+}
+
+
+
+std::string sinsp_filter_check_fspath::extract_fspath(sinsp_evt* evt, filtercheck_map_t& checks)
+{
+	std::string ret;
+	std::vector<extract_value_t> values;
+	sinsp_evt* extract_evt = evt;
+
+	auto it = checks.find(extract_evt->get_type());
+	if(it == checks.end())
+	{
+		return "";
```

Comment:
Can't we just return a const char* at this point, and return NULL? This would allow proper null values checking on the invoker, plus avoid the string copy below.

Reply:
See above I changed it to return true/false and fill in a vector of pointers, similar to how `sinsp_filter_check::extract()` (the variant that works on a vector of extracted values) works.